+nosnoc/+discrete_time_problem/Gcs.m

```diff
@@ -0,0 +1,870 @@
+            lambda_0 = obj.w.lambda(0,0,opts.n_s);
+            
+            obj.g.z(0,0,opts.n_s) = {dcs.g_z_fun(x_0, z_0, obj.w.u(1), v_global, [p_global;obj.p.p_time_var(1)])};
+            obj.g.c_lb(0,0,opts.n_s) = {dcs.c_fun(x_0, z_0, v_global, [p_global;obj.p.p_time_var(1)]), 0, inf};
+            
+            x_prev = obj.w.x(0,0,opts.n_s);
+            for ii=1:opts.N_stages
+                h0 = obj.p.T().val/(opts.N_stages*opts.N_finite_elements(ii));
+                
+                ui = obj.w.u(ii);
+                p_stage = obj.p.p_time_var(ii);
+                p = [p_global;p_stage];
+                if obj.opts.use_speed_of_time_variables && opts.local_speed_of_time_variable
+                    s_sot = obj.w.sot(ii);
+                elseif obj.opts.use_speed_of_time_variables
+                    s_sot = obj.w.sot();
+                else
+                    s_sot = 1;
+                end
+                if opts.time_optimal_problem && ~opts.use_speed_of_time_variables
+                    t_stage = obj.w.T_final()/(opts.N_stages*opts.N_finite_elements(ii));
+                elseif opts.time_optimal_problem
+                    t_stage = s_sot*obj.p.T()/opts.N_stages;
+                else
+                    t_stage = obj.p.T()/opts.N_stages;
+                end
+
+                sum_h = 0;
+                for jj=1:opts.N_finite_elements(ii)
+                    if obj.opts.use_fesd
+                        h = obj.w.h(ii,jj);
+                        sum_h = sum_h + h;
+                    elseif opts.time_optimal_problem && ~opts.use_speed_of_time_variables
+                        h = obj.w.T_final()/(opts.N_stages*opts.N_finite_elements(ii));
+                    else
+                        h = h0;
+                    end
+                    switch opts.rk_representation
+                      case RKRepresentation.integral
+                        % In integral representation stage variables are states.
+                        x_ij_end = opts.D_rk(1)*x_prev;
+                        for kk=1:opts.n_s
+                            x_ijk = obj.w.x(ii,jj,kk);
+                            z_ijk = obj.w.z(ii,jj,kk);
+                            lambda_ijk = obj.w.lambda(ii,jj,kk);
+
+                            fj = s_sot*dcs.f_x_fun(x_ijk, z_ijk, lambda_ijk, ui, v_global, p);
+                            qj = s_sot*dcs.f_q_fun(x_ijk, z_ijk, lambda_ijk, ui, v_global, p);
+                            xk = opts.C_rk(1, kk+1) * x_prev;
+                            for rr=1:opts.n_s
+                                x_ijr = obj.w.x(ii,jj,rr);
+                                xk = xk + opts.C_rk(rr+1, kk+1) * x_ijr;
+                            end
+                            obj.g.dynamics(ii,jj,kk) = {h * fj - xk};
+                            obj.g.z(ii,jj,kk) = {dcs.g_z_fun(x_ijk, z_ijk, ui, v_global, p)};
+                            obj.g.c_lb(ii,jj,kk) = {dcs.c_fun(x_ijk, z_ijk, v_global, p), 0, inf};
```

Comment:
once again, i vote for a lifted version of $c$ if it is nonlinear (maybe to have it even as default setting - this would require some testing)

Reply:
Yep  this is now implemented. I am happy to make it a default as it does actually seem to improve convergence a bit :)

Reply:
Lifted default is our friend (although one might check offline is it affine, this is the only case where lifting is an overkill).